jobsdb/jobsdb.go

```diff
@@ -1771,6 +1784,74 @@ func (jd *HandleT) invalidateCacheForJobs(ds dataSetT, jobList []*JobT) {
 	}
 }
 
+type moreToken struct {
+	retryAfterJobID       *int64
+	waitingAfterJobID     *int64
+	unprocessedAfterJobID *int64
+}
+
+func (jd *HandleT) GetToProcess(ctx context.Context, params GetQueryParamsT, more MoreToken) (*MoreJobsResult, error) { // skipcq: CRT-P0003
+
+	mtoken := &moreToken{}
+	if more != nil {
+		var ok bool
+		if mtoken, ok = more.(*moreToken); !ok {
+			return nil, fmt.Errorf("invalid token: %+v", more)
+		}
+	}
+	updateParams := func(params *GetQueryParamsT, jobs JobsResult, nextAfterJobID *int64) {
+		params.JobsLimit -= len(jobs.Jobs)
+		if params.EventsLimit > 0 {
+			params.EventsLimit -= jobs.EventsCount
+		}
+		if params.PayloadSizeLimit > 0 {
+			params.PayloadSizeLimit -= jobs.PayloadSize
+		}
+		params.AfterJobID = nextAfterJobID
+	}
+	var list []*JobT
+	params.AfterJobID = mtoken.retryAfterJobID
+	toRetry, err := jd.GetToRetry(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+	if len(toRetry.Jobs) > 0 {
+		retryAfterJobID := toRetry.Jobs[len(toRetry.Jobs)-1].JobID
+		mtoken.retryAfterJobID = &retryAfterJobID
+	}
+
+	list = append(list, toRetry.Jobs...)
+	if toRetry.LimitsReached {
+		return &MoreJobsResult{JobsResult: JobsResult{Jobs: list, LimitsReached: true}, More: mtoken}, nil
+	}
+	updateParams(&params, toRetry, mtoken.waitingAfterJobID)
+
+	waiting, err := jd.GetWaiting(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+	if len(waiting.Jobs) > 0 {
+		waitingAfterJobID := waiting.Jobs[len(waiting.Jobs)-1].JobID
+		mtoken.waitingAfterJobID = &waitingAfterJobID
+	}
+	list = append(list, waiting.Jobs...)
+	if waiting.LimitsReached {
+		return &MoreJobsResult{JobsResult: JobsResult{Jobs: list, LimitsReached: true}, More: mtoken}, nil
+	}
+	updateParams(&params, waiting, mtoken.unprocessedAfterJobID)
+
+	unprocessed, err := jd.GetUnprocessed(ctx, params)
+	if err != nil {
+		return nil, err
+	}
+	if len(unprocessed.Jobs) > 0 {
+		unprocessedAfterJobID := unprocessed.Jobs[len(unprocessed.Jobs)-1].JobID
+		mtoken.unprocessedAfterJobID = &unprocessedAfterJobID
+	}
+	list = append(list, unprocessed.Jobs...)
```

Comment:
[question] does it make sense to move this in the previous if block? what purpose does it serve if `len(unprocessed.Jobs) == 0`?